### ISSUE TITLE
ReLU activation with IndyLSTMCell

### ISSUE THREAD

**Issue by bshao001** [batzner/indrnn#19]
Sorry to open an issue here. However, I think that you are an expert on this topic, and IndRNNCell related upgrade in TensorFlow 1.10.0 might also be created by you.

I am trying to apply relu activation to IndyLSTMCell in the new TensorFlow (1.10). However, the loss becomes NAN after I make that change. The default activation tanh works for this cell. IndyGRUCell has the same problem. For IndRNNCell, both tanh and relu work. However, when I stack it to multiple layer, I did not see any performance (capacity) increase of the model (concerning the loss decrease speed over training epochs).

Can you please give me a hint to address this? Any suggestions would be very appreciated. Thanks!

**Comment by Sunnydreamrain** [comment]
Hi,
You need to add batch normalization layer to make IndRNN work. Following are two examples shown in pytorch. But I haven't done the tensorflow version yet. You can wait for Batzner's response.   
https://github.com/StefOe/indrnn-pytorch/blob/master/indrnn.py#L256  
https://github.com/Sunnydreamrain/IndRNN_pytorch/blob/master/cPTB/language_model.py#L127  
Thanks.

**Comment by Sunnydreamrain** [comment]
No. I did not do NMT experiments.
It is interesting to know that the current RNN models cannot work on the above simple problem. If it is because of not able to keep long-term memory, IndRNN can definitely work better as it can keep longer memory.
For now, I don't think it is necessary to use IndLSTM or IndGRU. One of the benefit introduced by IndRNN is the robust use of ReLU. For LSTM and GRU, the gate parameters should use sigmoid to produce an appropriate range to keep information flow. Moreover, there are much more parameters and computations in IndLSTM and IndGRU. I did do one experiment using IndLSTM at the very beginning, but as far as I recall, the performance is not significantly better than IndRNN.

Thanks.

**Comment by bshao001** [comment]
Thank you again for the reply. I am able to use relu for IndRNN without any problem (two layers). However, when I stack it to 3 layers, 4 layers, or more (I tried 3, 4, 6, 8 layers), I did not see any increase of model capacity. The model capacity is also very important to me, as I need to handle a large amount of data. Do you have any suggestions for that? Thank you again. If possible, please add me to your wechat, my ID is: bshao001_miami

**Comment by batzner** [comment]
For now, I think we can close this issue. Feel free to re-open it if any questions or issues come up.
